File: StatedSprite.java
import java.applet.Applet;
import java.awt.Graphics;
import java.awt.Image;

public class StatedSprite extends Sprite {
  protected Applet App = Utils.GetApp();
  
  ImgObj[] StateSprites;
  
  ImgObj TempStateSprite;
  
  protected int CurrentState;
  
  protected int NextState;
  
  protected int DefaultState;
  
  protected int NumStates;
  
  protected int CurrentFrame = 0;
  
  public int Action;
  
  int FrameCount = 0;
  
  public StatedSprite(int paramInt1, int paramInt2, int paramInt3, int paramInt4) {
    super(paramInt1, paramInt2, paramInt3, paramInt4);
  }
  
  public void InitMain(int paramInt1, int paramInt2) {
    this.NumStates = paramInt1;
    this.CurrentState = paramInt2;
    this.DefaultState = paramInt2;
    this.NextState = paramInt2;
    this.StateSprites = new ImgObj[paramInt1];
  }
  
  public void InitState(int paramInt1, int paramInt2, int paramInt3) {
    if (paramInt2 > 0)
      this.StateSprites[paramInt1] = new ImgObj(paramInt2, paramInt3); 
  }
  
  public void LoadImage(Image paramImage, int paramInt1, int paramInt2) {
    (this.StateSprites[paramInt2]).SpriteImage[paramInt1] = paramImage;
  }
  
  public void LoadImages(ImageExtractor paramImageExtractor, int paramInt1, int paramInt2) {
    for (byte b = 0; b < (this.StateSprites[paramInt1]).Frames; b++)
      (this.StateSprites[paramInt1]).SpriteImage[b] = paramImageExtractor.ExtractImage(b * this.width, paramInt2, this.width, this.height); 
  }
  
  public final void Paint(Graphics paramGraphics) {
    if (this.visible)
      paramGraphics.drawImage((this.StateSprites[this.CurrentState]).SpriteImage[this.CurrentFrame], this.xpos, this.ypos, this.App); 
  }
  
  public void SetNewState(int paramInt) {
    this.CurrentFrame = 0;
    this.FrameCount = 0;
    this.CurrentState = paramInt;
  }
  
  public final void Update() {
    this.TempStateSprite = this.StateSprites[this.CurrentState];
    if (this.TempStateSprite.FrameDelay != -1) {
      this.FrameCount++;
      if (this.FrameCount >= this.TempStateSprite.FrameDelay) {
        this.FrameCount = 0;
        this.CurrentFrame++;
        if (this.CurrentFrame == this.TempStateSprite.Frames) {
          this.CurrentFrame = 0;
          this.Action = this.CurrentState;
          this.CurrentState = this.NextState;
          if (this.DefaultState != -1)
            this.NextState = this.DefaultState; 
        } 
      } 
    } 
  }
}


/* Location:              C:\Users\brody\Downloads\invader.jar!\StatedSprite.class
 * Java compiler version: 1 (45.3)
 * JD-Core Version:       1.1.3
 */

File: web/js/StatedSprite.js
// StatedSprite.js — port of StatedSprite.java
// Extends Sprite with multi-state sprite animation.

import { Sprite } from './Sprite.js';
import { ImgObj } from './ImgObj.js';

export class StatedSprite extends Sprite {
  constructor(x, y, w, h) {
    super(x, y, w, h);
    this.stateSprites   = null;   // ImgObj[]
    this.currentState   = 0;
    this.nextState      = 0;
    this.defaultState   = 0;
    this.numStates      = 0;
    this.currentFrame   = 0;
    this.action         = 0;
    this.frameCount     = 0;
  }

  initMain(numStates, defaultState) {
    this.numStates    = numStates;
    this.currentState = defaultState;
    this.defaultState = defaultState;
    this.nextState    = defaultState;
    this.stateSprites = new Array(numStates).fill(null);
  }

  initState(stateIdx, frames, frameDelay) {
    if (frames > 0) {
      this.stateSprites[stateIdx] = new ImgObj(frames, frameDelay);
    }
  }

  loadImage(img, frameIdx, stateIdx) {
    this.stateSprites[stateIdx].spriteImage[frameIdx] = img;
  }

  loadImages(extractor, stateIdx, y) {
    const s = this.stateSprites[stateIdx];
    for (let i = 0; i < s.frames; i++) {
      s.spriteImage[i] = extractor.extractImage(i * this.width, y, this.width, this.height);
    }
  }

  setNewState(state) {
    this.currentFrame = 0;
    this.frameCount   = 0;
    this.currentState = state;
  }

  paint(ctx) {
    if (!this.visible) return;
    const s = this.stateSprites[this.currentState];
    if (!s) return;
    const img = s.spriteImage[this.currentFrame];
    if (img) ctx.drawImage(img, this.xpos, this.ypos);
  }

  update() {
    const s = this.stateSprites[this.currentState];
    if (!s || s.frameDelay === -1) return;
    this.frameCount++;
    if (this.frameCount >= s.frameDelay) {
      this.frameCount = 0;
      this.currentFrame++;
      if (this.currentFrame === s.frames) {
        this.currentFrame = 0;
        this.action       = this.currentState;
        this.currentState = this.nextState;
        if (this.defaultState !== -1) this.nextState = this.defaultState;
      }
    }
  }
}
